src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs

```diff
@@ -34,27 +35,76 @@ public static ImmutableQueue<T> Create<T>()
         [Pure]
         public static ImmutableQueue<T> Create<T>(T item)
         {
-            return ImmutableQueue<T>.Empty.Enqueue(item);
+            return new ImmutableQueue<T>(ImmutableStack.Create(item), ImmutableStack<T>.Empty);
         }
 
         /// <summary>
-        /// Creates a new immutable collection prefilled with the specified items.
+        /// Creates a new immutable queue from the specified items.
         /// </summary>
-        /// <typeparam name="T">The type of items stored by the collection.</typeparam>
-        /// <param name="items">The items to prepopulate.</param>
-        /// <returns>The new immutable collection.</returns>
+        /// <typeparam name="T">The type of items to store in the queue.</typeparam>
+        /// <param name="items">The enumerable to copy items from.</param>
+        /// <returns>The new immutable queue.</returns>
         [Pure]
         public static ImmutableQueue<T> CreateRange<T>(IEnumerable<T> items)
         {
             Requires.NotNull(items, nameof(items));
 
-            var queue = ImmutableQueue<T>.Empty;
-            foreach (var item in items)
+            var ilist = items as IList<T>;
+            if (ilist != null)
             {
-                queue = queue.Enqueue(item);
+                return CreateRange(ilist: ilist);
             }
 
-            return queue;
+            using (IEnumerator<T> e = items.GetEnumerator())
+            {
+                if (!e.MoveNext())
+                {
+                    return ImmutableQueue<T>.Empty;
+                }
+
+                var forwards = ImmutableStack.Create(e.Current);
```

Comment:
@ianhays We try to defer the allocation where we [reverse the backwards stack](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs#L199) as much as possible. Pushing 1 item onto the forwards stack makes that take 2 Dequeues to happen instead of 1. (This is basically the existing behavior because [we push once](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs#L154) onto the forwards stack if we're empty.)